ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -262,76 +263,85 @@ chainSyncClient _tracer cfg btime (ClockSkew maxSkew) getCurrentChain
 
       -- Get the 'HeaderState' at genesis (0).
       let candidateChain' = Empty genesisPoint
-          candidateHeaderState' =
-            getHeaderStateFor curLedger candidateChain candidateChain'
+
+      candidateChainState' <- case rewindChainState cfg curChainState (SlotNo 0) of
+        Nothing -> disconnect $ ForkTooDeep genesisPoint theirHead
+        Just c -> pure c
 
       writeTVar varCandidate CandidateState
         { candidateChain       = candidateChain'
-        , candidateHeaderState = candidateHeaderState'
+        , candidateChainState  = candidateChainState'
         }
       return $ requestNext varCandidate
 
-    requestNext :: TVar m (CandidateState blk hdr)
+    requestNext :: TVar m (CandidateState hdr)
                 -> Consensus ClientStIdle hdr m
     requestNext varCandidate = SendMsgRequestNext
       (handleNext varCandidate)
       (return (handleNext varCandidate)) -- when we have to wait
 
-    handleNext :: TVar m (CandidateState blk hdr)
+    handleNext :: TVar m (CandidateState hdr)
                -> Consensus ClientStNext hdr m
     handleNext varCandidate = ClientStNext
       { recvMsgRollForward  = ChainSyncClient .: rollForward  varCandidate
       , recvMsgRollBackward = ChainSyncClient .: rollBackward varCandidate
       }
 
-    rollForward :: TVar m (CandidateState blk hdr)
+    rollForward :: TVar m (CandidateState hdr)
                 -> hdr -> Point hdr
                 -> m (Consensus ClientStIdle hdr m)
     rollForward varCandidate hdr theirHead = atomically $ do
       currentSlot <- getCurrentSlot btime
+      -- TODO #472
+      curLedger <- ledgerState <$> getCurrentLedger
       let theirSlot = AF.pointSlot theirHead
 
       when (unSlotNo theirSlot > unSlotNo currentSlot + maxSkew) $
         disconnect $ TooFarInTheFuture theirSlot currentSlot
 
-      -- TODO #472
-      curLedger <- ledgerState <$> getCurrentLedger
-
       CandidateState {..} <- readTVar varCandidate
 
-      candidateHeaderState' <-
-        case runExcept $ advanceHeader curLedger hdr candidateHeaderState of
-          Left ledgerError            -> disconnect $ LedgerError ledgerError
-          Right candidateHeaderState' -> return candidateHeaderState'
+      ledgerView <-
```

Comment:
Would it make sense to maintain this `ledgerView` in the `CandidateState`? We can then use `atSlot`to check whether it is still valid when rolling forward.